misc/peer-store/src/memory_store.rs

```diff
@@ -52,67 +72,70 @@ impl<T> MemoryStore<T> {
     /// [`FromSwarm::NewExternalAddrOfPeer`] to the swarm, e.g. via
     /// [`Swarm::add_peer_address`](libp2p_swarm::Swarm::add_peer_address).
     ///
     /// Returns `true` if the address is new.
-    pub fn update_address(&mut self, peer: &PeerId, address: &Multiaddr) -> bool {
-        let is_updated = self.update_address_silent(peer, address, true);
-        if is_updated {
-            self.push_event_and_wake(crate::store::Event::RecordUpdated(*peer));
-        }
-        is_updated
+    pub fn add_address(&mut self, peer: &PeerId, address: &Multiaddr) -> bool {
+        self.add_address_inner(peer, address, true)
     }
 
-    /// Update an address record without notifying swarm.
+    /// Update an address record and notify the swarm.
     ///
     /// Returns `true` if the address is new.
-    fn update_address_silent(
+    fn add_address_inner(
         &mut self,
         peer: &PeerId,
         address: &Multiaddr,
-        permanent: bool,
+        is_permanent: bool,
     ) -> bool {
-        if let Some(record) = self.records.get_mut(peer) {
-            return record.update_address(address, permanent);
+        let record = self
+            .records
+            .entry(*peer)
+            .or_insert(PeerRecord::new(self.config.record_capacity));
+        let is_new = record.add_address(address, is_permanent);
+        if is_new {
+            self.push_event_and_wake(Event::PeerAddressAdded {
+                peer_id: *peer,
+                address: address.clone(),
+                is_permanent,
+            });
         }
-        let mut new_record = PeerRecord::new(self.config.record_capacity);
-        new_record.update_address(address, permanent);
-        self.records.insert(*peer, new_record);
-        true
+        is_new
     }
 
     /// Remove an address record.
     ///
     /// Returns `true` when the address existed.
     pub fn remove_address(&mut self, peer: &PeerId, address: &Multiaddr) -> bool {
-        let is_updated = self.remove_address_silent(peer, address, true);
-        if is_updated {
-            self.push_event_and_wake(crate::store::Event::RecordUpdated(*peer));
-        }
-        is_updated
+        self.remove_address_inner(peer, address, true)
     }
 
-    /// Remove an address record without notifying swarm.
+    /// Remove an address record and notify the swarm.
     ///
     /// Returns `true` when the address is removed, `false` if the address didn't exist
     /// or the address is permanent and `force` false.
-    fn remove_address_silent(&mut self, peer: &PeerId, address: &Multiaddr, force: bool) -> bool {
+    fn remove_address_inner(&mut self, peer: &PeerId, address: &Multiaddr, force: bool) -> bool {
         if let Some(record) = self.records.get_mut(peer) {
             if record.remove_address(address, force) {
-                if record.addresses.is_empty() && record.custom_data.is_none() {
+                if record.addresses.is_empty() {
```

Comment:
Are we not considering `custom_data` when removing the entire entry?

Reply:
Ah yes you're right of course!

```suggestion
                if record.addresses.is_empty() && record.custom_data.is_none() {
```